### PR TITLE
Fabrica improvements 4

### PR DESCRIPTION
**Pull request by bklaing2**
## Purpose
Update to Format component for DOIs to suggest current MIME format types.
[Vercel Build](https://bracco-ort38lrhx-datacite.vercel.app/)

closes: https://github.com/datacite/datacite/issues/1515

## Approach
It utilizes the [mime/lite](https://www.npmjs.com/package/mime) package for getting all of the MIME types. These are then suggested to the user in a dropdown as they type.

image

#### Open Questions and Pre-Merge TODOs
- [ ] Currently, the component only suggests the MIME types (application/pdf, text/xml,
video/mpeg) as opposed to the file extension (PDF, XML, MPG). Should it also suggest the extensions themselves, or just suggest the extensions instead of the MIME types?

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)

- [x] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Reviewer, please remember our [guidelines](https://datacite.atlassian.net/wiki/spaces/TEC/pages/1168375809/Pull+Request+Guidelines):

- Be humble in the language and feedback you give, ask don't tell.
- Consider using positive language as opposed to neutral when offering feedback. This is to avoid the negative bias that can occur with neutral language appearing negative.
- Offer suggestions on how to improve code e.g. simplification or expanding clarity.
- Ensure you give reasons for the changes you are proposing.

### CODE REVIEW

**Review by svogt0511 — COMMENTED** [review]
I tested in the vercel preview at https://bracco-j2a1okxv6-datacite.vercel.app/.  It shows the suggestions but there are still issues.

I have an existing doi at: https://bracco-j2a1okxv6-datacite.vercel.app/dois/10.0309%2Fformat-test-1.  It has 2 formats: application/pdf and application/pdf+jpg.  The second format is made up. When I edit that form, it indicates there are 2 formats, but their values don't show.  If I try to add a format (standard or made up) it doesn't get saved.

Same in the create doi form, it shows the list of IANA formats, but if you choose one, it doesn't save it.  Additionally, you can't create your own.

Let me know when this is fixed and I will review again.

Additionally, I got a weird warning in the console, something about invalid json and it printed out a document type definition (<!DOCTYPE...), but unfortunately I can't reproduce it. It may be unrelated.

Thanks.

**Comment by bklaing2** [comment]
@svogt0511 I've fixed the issues you addressed, and deployed a new version with the changes. Here is the [Vercel build](https://bracco-ort38lrhx-datacite.vercel.app/). Please let me know if there is anything else!

File: app/components/doi-format.js [hunk 1]
```diff
@@ -1,16 +1,59 @@
...
+          this.set('formats', FORMATS);
+        }
+      }
+    },
+    searchFormat(query) {
+      this.set('formats', getMatchingFormats(query))
+    },
+    selectFormat(formatExtension) {
+      this.model.get('formats').replace(this.index, 1, [ formatExtension ]);
+      // this.set('fragment', formatExtension);
```
Comment (svogt0511) [inline app/components/doi-format.js:48]: Remove commented code.

File: app/components/doi-format.js [hunk 2]
```diff
@@ -1,16 +1,59 @@
...
     },
-    selectFormat() {
-      this.model.get('formats').replace(this.index, 1, [ this.fragment ]);
+    onBlur() {
+      console.log(0)
+      // this.model.get('formats').replace(this.index, 1, [ this.fragment ]);
```
Comment (svogt0511) [inline app/components/doi-format.js:52]: Remove commented code.

File: app/components/doi-format.js [hunk 3]
```diff
@@ -1,16 +1,59 @@
...
+        e.keyCode === 13 &&
+        select.isOpen &&
+        !select.highlighted &&
+        !isBlank(select.searchText)
+      ) {
+        if (!this.selected.includes(select.searchText)) {
+          this.formats.push(select.searchText);
+          select.actions.choose(select.searchText);
+          this.model.get('formats').replace(this.index, 1, [ select.searchText ]);
+          // this.set('fragment', select.searchText);
```
Comment (svogt0511) [inline app/components/doi-format.js:38]: Remove commented code.

Comment (svogt0511) [inline app/components/doi-format.js:38]: Remove commented code.

File: app/components/doi-format.js [hunk 4]
```diff
@@ -1,16 +1,59 @@
...
+    selectFormat(formatExtension) {
+      this.model.get('formats').replace(this.index, 1, [ formatExtension ]);
+      // this.set('fragment', formatExtension);
     },
-    selectFormat() {
-      this.model.get('formats').replace(this.index, 1, [ this.fragment ]);
+    onBlur() {
+      console.log(0)
+      // this.model.get('formats').replace(this.index, 1, [ this.fragment ]);
+      console.log(9)
```
Comment (svogt0511) [inline app/components/doi-format.js:53]: Remove console.log.

File: app/components/doi-format.js [hunk 5]
```diff
@@ -1,16 +1,59 @@
...
+      this.set('formats', getMatchingFormats(query))
+    },
+    selectFormat(formatExtension) {
+      this.model.get('formats').replace(this.index, 1, [ formatExtension ]);
+      // this.set('fragment', formatExtension);
     },
-    selectFormat() {
-      this.model.get('formats').replace(this.index, 1, [ this.fragment ]);
+    onBlur() {
+      console.log(0)
```
Comment (svogt0511) [inline app/components/doi-format.js:51]: Remove console.log.

**Review by svogt0511 — COMMENTED** [review]
Looks good.  Seems to work. However, there is still one tiny issue.  Kristian specified that the 'name' of the selected value should be displayed in the field once the user picks from the IANA list.  I think this is easily done with a function that picks off of the string past the last '/' in the media type past ('application/pdf' would display 'pdf').  We can ask Kristian tomorrow in the dev meeting.  I actually think it is ok the way it is.  I think it is confusing to display the name only. 

Regarding your comment:

> Currently, the component only suggests the MIME types (application/pdf, text/xml,
video/mpeg) as opposed to the file extension (PDF, XML, MPG). Should it also suggest the extensions themselves, or just suggest the extensions instead of the MIME types?

I would go with mime types and not file extensions.  We can talk about that, as well.

**Review by svogt0511 — CHANGES_REQUESTED** [review]
I meant to mark the comments need to be addressed.

File: app/components/doi-format.js (hunk 2 above)
Comment (bklaing2) [reply app/components/doi-format.js]: Removed this

File: app/components/doi-format.js (hunk 3 above)
Comment (bklaing2) [reply app/components/doi-format.js]: Removed this

Comment (bklaing2) [reply app/components/doi-format.js]: Removed this

File: app/components/doi-format.js (hunk 4 above)
Comment (bklaing2) [reply app/components/doi-format.js]: Removed this

File: app/components/doi-format.js (hunk 5 above)
Comment (bklaing2) [reply app/components/doi-format.js]: Removed this

**Review by svogt0511 — APPROVED** [review]
Looks good to me.

**Comment by svogt0511** [comment]
@kjgarza I am changing the review to 'Approved'.  This does 99.9% of what is stated in the spec except for the minor issue of changing the display of the selected format to the IANA format 'Name' instead of the 'Template'. 

At present, I don't see an option with ember-power-select that will allow us to do this, and trying to work around the ember-power-select implementation is not a viable option.  @bklaing2  and I spent a fair amount of time looking for a workaround.

I think we should consider this task done.  We can add this task as a github issue that can be taken care of at a later date.

In the meantime, you can review this and the rest of the project work in [Vercel](https://bracco-9ud7n9kdr-datacite.vercel.app/).

**Comment by svogt0511** [comment]
@kjgarza And, it looks like ember-power-select DOES have an option for customizing the display of the selected item and it DOES seem to work, so we should be able to do this as you specified in the product spec.

**Pull request merged**